pilot/pkg/xds/ep_filters.go

```diff
@@ -111,10 +111,17 @@ func EndpointsByNetworkFilter(push *model.PushContext, proxyNetwork string, endp
 				}
 				// TODO: figure out a way to extract locality data from the gateway public endpoints in meshNetworks
 				gwEp.Metadata = util.BuildLbEndpointMetadata(network, model.IstioMutualTLSModeLabel)
-				lbEndpoints = append(lbEndpoints, gwEp)
+				remoteLbEps = append(remoteLbEps, gwEp)
 			}
 		}
 
+		// Only include remote endpoints for an ingress-gateway if we have no local endpoints
+		// Normal proxies or egress gateways always get all the endpoints.
+		isIngress := b.proxyType == model.Router && b.routerMode == model.SniDnatRouter
```

Comment:
Can you confirm `SniDnatRouter` is a valid way to determine that this is ingress? @hzxuzhonghu

Reply:
Not right, the sni-dnat is for all the gateways as you can see from the default yaml

Reply:
Is the label `istio=ingressgateway` reliable? Seems like the user really doesn't have to have this and it could me missing if they have a custom setup.

Reply:
No, we cannot rely on the labels. The goal is to determine if its an ingress or egress gateway? We don't have a way to do this - egress and ingress gateways are technically the same and are just used different as a standard practice.

Reply:
In comment thread above @hzxuzhonghu explained a case this would break if we are going from `client in network a -> egress in network a -> ingress in network b -> server in network -> b`, where we have the destination service in both networks. So we'd still want egress to talk to ingress, but ingress should only go to the actual endpoints. 

Alternatives would be:
1. wait for BTS to allow header based routing and `Forwarded` header. 
2. Use priority + outlier detection to just use the other network as a failover in the egress case. 
3. Something at the user-config level that's only applied to ingress
4. Ignore that edge-case

Reply:
What about setting `ISTIO_META_REQUESTED_NETWORK_VIEW` on the gateway Deployment? It appears to be intended for this purpose, of filtering out endpoints in other networks. I'm not sure if it works though

Reply:
I don't see any references to that outside of YAML :/  Do you see references in docs?

Reply:
I'm assuming you mean, if it's not "external" then don't build the cross-network endpoints, right? If we go that direction, and we already as default for egress, we'd need to make sure it's documented for people who want to setup custom gateways.

Reply:
I saw it when looking at code; it sets node metadata `RequestedNetworkView` which is used to filter out endpoints in clusters https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L219

It seems like if the value is the name of the gateway's network, it would filter out endpoints from other networks

Reply:
My bad - grep shouldn't have included `ISTIO_META_`. Seems like that would work. For the case where we have a service in network b, it wouldn't be reachable from `network a` ingress, but that's probably fine if we're only setting the network-view for a special cross-cluster ingress.

Reply:
`RequestedNetworkView`  could be a good direction, And I find that we donot consider it when filtering eds currently.

Reply:
Seems like a good idea, but we should make sure we don't add additional confusion to multi-network installation. 

1. The default for all proxies is currently `{"":true}` - meaning by default, proxies are _already_ filtering out DNSLB services for cross-network. We should change the default to be "this proxy sees all networks". 

2. Users shouldn't really use the same gateway for normal ingress and cross-network. Since the cross-network concern has special requirements, does it make sense to have a `istio-cross-network` deployment template to help users set things up, and provide an example for when the want to do something custom. 

cc @nmittler

Reply:
Yes, we should have completely different ingress deployments for an ingress used for east-west traffic in general. We should also update the samples to use custom ingresses:

- https://github.com/istio/istio/tree/master/samples/cross-network-gateway
- https://github.com/istio/istio/tree/master/samples/istiod-gateway

WDYT?